### ISSUE TITLE
Backend for MA Dot2

### ISSUE THREAD

**Issue by Spacelord09** [cbdevnet/midimonster#21]
Can anyone add a Backend that uses the Webremote from the Dot2?

**Comment by cbdevnet** [comment]
Yes, I'm currently in the process of developing that (albeit with a focus on the GrandMA2 Web remote - though I think they use the same commands).

**Comment by Spacelord09** [comment]
@cbdevnet Can you make a documentation for it?

**Comment by cbdevnet** [comment]
Yes I'll publish it with documentation as always, but it'll take at least a day more to be somewhat usable.

**Comment by Spacelord09** [comment]
@cbdevnet Can you give me an update?

**Comment by cbdevnet** [comment]
@Spacelord09 I've just pushed a preliminary version containing my current development status. It's not quite finished yet and has some limitations:

* It outputs a lot of debug information
* It currently is write-only, channel events are only sent to the MA, not consumed by it
* Fader executors (and their buttons) seem to work, I haven't tested button executors yet.
* Command line events are sent, but I'm not sure they're being handled by the MA yet - I think I'm missing some piece of communication there
* I have so far only tested it with GradMA2 OnPC, but I have confirmed that the Dot2 at least has the same initial exchange. Maybe you can provide some feedback on how well it works currently :)

You can build it with `make maweb.so` in the `backends/` directory - once it's stable I'll add it to the main set of backends built automatically.

**Comment by Spacelord09** [comment]
@cbdevnet I can test it for you.

**Comment by Spacelord09** [comment]
Hey @cbdevnet That wont work with the Dot2.. I tested it only with the Dot2.


`root@MIDIMONSTER:/midimonster# ./midimonster /midimonster/configs/TEST#1.cfg
Registered backend maweb
Registered backend osc
Registered backend sacn
Registered backend midi
MIDI client ID is 128
Registered backend lua
Registered backend artnet
Registered backend loopback
Registered backend evdev
Created maweb instance mw1
maweb backend registering 1 descriptors to core
Skipping start of backend osc
Skipping start of backend sacn
Skipping start of backend midi
Skipping start of backend lua
Skipping start of backend artnet
Skipping start of backend loopback
Skipping start of backend evdev
maweb message (43): {"status":"server ready", "appType":"dot2"}
maweb connection established
maweb message (64): {"realtime":false,"session":11,"forceLogin":true,"worldIndex":0}
maweb session id is now 11
maweb sending user credentials
maweb message (118): {"prompt":" [Fixture]>","promptcolor":"#808080","realtime":false,"responseType":"login","result":false,"worldIndex":0}
maweb login failed
`

**Comment by Spacelord09** [comment]
@cbdevnet can you make a sample config?

**Comment by cbdevnet** [comment]
@Spacelord09 It seems that the dot2 does not have full user management, instead you can just set the password for the web remote - the username is always 'remote'. I'll add it to the documentation.

Example configuration (I use it to test the maweb backend with my BCF fader wing in `NC C` mode):

```
[backend midi]
detect = on

[midi bcf]
read = BCF
write = BCF

[maweb ma]
host = 10.23.23.248
user = web
password = web

[map]
bcf.channel{0..7}.pitch > bcf.channel{0..7}.pitch
bcf.channel{0..7}.pitch > ma.page1.fader{1..8}
bcf.channel0.note{16..23} > ma.page1.upper{1..8}
bcf.channel0.note{24..31} > ma.page1.lower{1..8}

**Comment by cbdevnet** [comment]
I modified the backend so it can detect whether it is connected to a dot2 and automatically set the user name to `remote`

**Comment by Spacelord09** [comment]
@cbdevnet I´ve tested it. It works!

But.... If I pressed a button mapped to " ma.page1.upper1" the key don't work...

**Comment by Spacelord09** [comment]
Can you add motorized fader support for BCF?

**Comment by cbdevnet** [comment]
You can use the motorfaders with the MIDIMonster by mapping the appropriate channels as output on the MIDI instance (the same one's the faders send as inputs).

I'm still looking into how to get useful fader data as output from the MA Web Remote.

I'll look into the way the executor buttons are adressed on the dot, I think they have only two buttons and a fader for fader execs, is that right?

**Comment by Spacelord09** [comment]
Yea. right!

**Comment by cbdevnet** [comment]
OK, I've analyzed the dot2 a bit further, the controls are numbered in a really weird way (right-to-left across the wings), I've added some bits so that in the current version, the `upper`/`lower` keys for fader execs on the dot should work, as well as being able to address all button executors on the console :)

**Comment by Spacelord09** [comment]
It´s working!

(Can we talk in german?? ;DDD)

**Comment by cbdevnet** [comment]
Ja prinzipiell schon ;)

**Comment by Spacelord09** [comment]
Jo machen wir so😄

**Comment by cbdevnet** [comment]
Barring some minor problems, this was implemented in b4b27aa4a90899d5b025bbef11d444d4c47800d9
